catalog/spatial/registry/registry-api-impl/src/main/java/org/codice/ddf/registry/api/impl/RegistryMetacardHandler.java

```diff
@@ -0,0 +1,145 @@
+import org.osgi.service.event.EventHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.impl.AttributeImpl;
+
+public class RegistryMetacardHandler implements EventHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RegistryMetacardHandler.class);
+
+    private static final int SHUTDOWN_TIMEOUT_SECONDS = 60;
+
+    private static final String METACARD_PROPERTY = "ddf.catalog.event.metacard";
+
+    private static final String CREATED_TOPIC = "ddf/catalog/event/CREATED";
+
+    private static final String UPDATED_TOPIC = "ddf/catalog/event/UPDATED";
+
+    private static final String DELETED_TOPIC = "ddf/catalog/event/DELETED";
+
+    private final ExecutorService executor;
+
+    private final FederationAdminService federationAdminService;
+
+    public RegistryMetacardHandler(ExecutorService executor,
+            FederationAdminService federationAdminService) {
+        this.executor = executor;
+        this.federationAdminService = federationAdminService;
+    }
+
+    @Override
+    public void handleEvent(Event event) {
+        Metacard mcard = (Metacard) event.getProperty(METACARD_PROPERTY);
+        if (mcard == null || !RegistryUtility.isInternalRegistryMetacard(mcard)) {
+            return;
+        }
+        executor.execute(() -> processEvent(mcard, event.getTopic()));
+    }
+
+    private void processEvent(Metacard mcard, String topic) {
+        try {
+            Security.runAsAdminWithException(() -> {
+                if (topic.equals(DELETED_TOPIC)) {
+                    processMetacardDelete(mcard);
+                } else if (topic.equals(CREATED_TOPIC) || topic.equals(UPDATED_TOPIC)) {
+                    processMetacardCreateUpdate(mcard);
+                }
+                return null;
+            });
+        } catch (PrivilegedActionException e) {
+            LOGGER.error("Error processing registry metacard event.", e);
+        }
+    }
+
+    private void processMetacardCreateUpdate(Metacard mcard) throws FederationAdminException {
+        List<Metacard> metacards = federationAdminService.getRegistryMetacardsByRegistryIds(
+                Collections.singletonList(RegistryUtility.getRegistryId(mcard)));
+        if (CollectionUtils.isEmpty(metacards)) {
+            prepareMetacard(mcard, null);
+            federationAdminService.addRegistryEntry(mcard);
+
+        } else if (shouldUpdate(mcard, metacards.get(0))) {
+            prepareMetacard(mcard,
+                    metacards.get(0)
+                            .getId());
+            federationAdminService.updateRegistryEntry(mcard);
+        }
+    }
+
+    private void processMetacardDelete(Metacard mcard) throws FederationAdminException {
+        List<Metacard> metacards = federationAdminService.getRegistryMetacardsByRegistryIds(
+                Collections.singletonList(RegistryUtility.getRegistryId(mcard)));
```

Comment:
We're calling getRegistryID on the same metacard a couple of times. Should we just save off the registryID?